### ISSUE TITLE
False positive ``unnecessary 'unsafe' block`` warning?

### ISSUE THREAD

**Issue by Vurv78** [rust-lang/rust-analyzer#9576]
I'm not sure if this is a part of ``GenericDetour::call`` being generated by a macro, I thought this was fixed with expanding macros though. Removing the unsafe block just causes a ``missing-unsafe`` error
```rs
use detour::GenericDetour;

type Add5 = extern "Rust" fn(i32) -> i32;
extern "Rust" fn add_5(n: i32) -> i32 { n + 5 }
extern "Rust" fn hookfn(n: i32) -> i32 { n + 50 }

fn main() -> Result<(), detour::Error> {
	let hook = unsafe { GenericDetour::<Add5>::new(add_5, hookfn)? };
	
	// This unsafe is needed (afaik) yet causes a ``unnecessary `unsafe` block`` diagnostic?
	let _res = unsafe { hook.call(50) }; 

	Ok(())
}
```

**Comment by flodiebold** [comment]
`missing-unsafe` is our diagnostic, "unnecessary unsafe block" comes from rustc. When in doubt, rustc is right ;) Apparently the `call` method is generated by a macro; depending on the macro (I didn't spend much time looking for the call), it might be necessary to enable `rust-analyzer.experimental.procAttrMacros`, or it might be another problem.

**Comment by bjorn3** [comment]
It is a decl macro: https://docs.rs/detour/0.8.0/src/detour/macros.rs.html#36-165

**Comment by flodiebold** [comment]
Oof. There are a lot of unsafe and safe versions of the method there, so RA might be getting the method selection wrong.

**Comment by Vurv78** [comment]
Oops, I didn't see the missing-unsafe error is from ra, but the warning is from rustc... So it should be safe but it is seen as unsafe by rust-analyzer

**Comment by rossmacarthur** [comment]
Encountered this when using `wasm-bindgen`

```rust
#[wasm_bindgen(module = "/index.js")]
extern "C" {
    #[wasm_bindgen]
    pub fn my_function() -> JsValue
}
```

`my_function` is safe here but ra thinks it is unsafe.

**Comment by lnicola** [comment]
@rossmacarthur try enabling `rust-analyzer.experimental.procAttrMacros`.

**Comment by nbars** [comment]
I believe I've got the same issue when using the `log` crates logging functions:
![Screenshot_2021-08-28_14-47-13](https://user-images.githubusercontent.com/1810786/131218629-a239fc1f-092d-4bfe-8951-0f229afd752c.png)

However, `cargo build` and `cargo check` work just fine. I'm using `rust-analyzer 996300f4a 2021-08-23 stable` and also tried the previous release.

Is there any way for me to get more details about why the rust-analyzer suspects the logging functions to be unsafe?

Keep up the good work :)

**Comment by nbars** [comment]
It appears that using the nightly channel via. `"rust-analyzer.updates.channel": "nightly"` solves the problem.

**Comment by lnicola** [comment]
@nbars see #10022

**Comment by yatesco** [comment]
> It appears that using the nightly channel via `"rust-analyzer.updates.channel": "nightly"` solves the problem.

I'm still getting this error, even with the nightly rust-analyzer (with stable rust and 2021 profile)

**Comment by ModProg** [comment]
> nevermind this might be an error with my vim config. VScode 
seems to work fine.
@yatesco if you still have the problem with vim, idk if you use lspconfig, but this is working for me with rust-tools (a wrapper around lspconfig)
<details>
  <summary>My Config</summary>

```lua
            require('rust-tools').setup {
                tools = {
                    autoSetHints = true,
                    hover_with_actions = true,
                    runnables = {use_telescope = true},
                    inlay_hints = {
                        show_parameter_hints = false,
                        parameter_hints_prefix = "",
                        other_hints_prefix = ""
                    }
                },
                -- see https://github.com/neovim/nvim-lspconfig/blob/master/CONFIG.md#rust_analyzer
                server = {
                    settings = {
                        -- https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/user/generated_config.adoc
                        ["rust-analyzer"] = {
                            checkOnSave = {command = "clippy"},
                            experimental = {procAttrMacros = true},
                            updates = {channel = "nightly"}
                        }
                    }
                }
            }
```
</details>

**Comment by yatesco** [comment]
Thanks @ModProg - it's working on the latest nightly now.

**Comment by zicklag** [comment]
For anybody else still running into this, disabling the `missing-unsafe` diagnostic might help:

```json
  "rust-analyzer.diagnostics.disabled": [
    "missing-unsafe"
  ],
```

**Comment by Veykril** [comment]
This doesn't happen for me anymore
